clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/BlobSnapshotManager.java

```diff
@@ -28,60 +30,65 @@
 
   /**
    * Constructor for the BlobSnapshotManager
    */
   public BlobSnapshotManager(
-      String basePath,
-      long snapshotRetentionTime,
-      ReadOnlyStoreRepository readOnlyStoreRepository) {
-    this.basePath = basePath;
-    this.snapshotRetentionTime = snapshotRetentionTime;
+      ReadOnlyStoreRepository readOnlyStoreRepository,
+      StorageEngineRepository storageEngineRepository) {
     this.readOnlyStoreRepository = readOnlyStoreRepository;
-    this.snapShotTimestamps = new VeniceConcurrentHashMap<>();
+    this.storageEngineRepository = storageEngineRepository;
+
     this.concurrentSnapshotUsers = new VeniceConcurrentHashMap<>();
   }
 
   /**
-   * Checks if the snapshot is stale, if it is and no one is using it, it updates the snapshot,
-   * otherwise it increases the count of people using the snapshot
+   * Recreate a snapshot for a hybrid store and return true if the snapshot is successfully created
+   * If the snapshot is being used by some other host, throw an exception
    */
-  public void maybeUpdateHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
-    if (rocksDB == null || topicName == null) {
-      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+  public boolean recreateSnapshotForHybrid(BlobTransferPayload payload) {
+    String topicName = payload.getTopicName();
+    String storeName = payload.getStoreName();
+    int partitionId = payload.getPartition();
+
+    if (!isStoreHybrid(storeName)) {
+      LOGGER.info("Store {} is not hybrid, skipping snapshot re-create", storeName);
+      return false;
     }
-    if (!isStoreHybrid(topicName)) {
-      LOGGER.warn("Store {} is not hybrid, skipping snapshot update", topicName);
-      return;
-    }
-    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
-    snapShotTimestamps.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
+
     concurrentSnapshotUsers.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
     concurrentSnapshotUsers.get(topicName).putIfAbsent(partitionId, new AtomicLong(0));
+
     try (AutoCloseableLock ignored = AutoCloseableLock.of(lock)) {
-      if (isSnapshotStale(topicName, partitionId)) {
-        if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
-            || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
-          updateHybridSnapshot(rocksDB, topicName, partitionId);
-          snapShotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
-        }
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
+          || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
+        createSnapshot(topicName, partitionId);
+      } else {
+        // the snapshot is being used, cannot generate a new snapshot at the same time
```

Comment:
why not just re-use the same snapshot?

Reply:
In hybrid stores, we do not allow multiple hosts to share the same snapshot. When the second GET request is received, the system checks the map and recognizes that one host is already utilizing that snapshot. However, the offset record will send to the second client (obtained before checking the snapshot's existence) may not correspond to the snapshot.

Reply:
Sync offline, we will (1) restrict the number of concurrent users, (2) allow the use of only non-stale snapshot within the set concurrency limit, and (3) monitor and record snapshot offsets and timestamps, updating them when generating a new snapshot to replace any stale ones.